### PR TITLE
MUL-288: CLI command to generate settlements

### PR DESCRIPTION
**Pull request by beniaminblaszkiewicz**
add new CLI command `bitbag:settlement:generate` to generate settlements.

### CODE REVIEW

File: spec/Component/Settlement/PeriodStrategy/MonthlySettlementPeriodResolverSpec.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+use PhpSpec\ObjectBehavior;
+
+final class MonthlySettlementPeriodResolverSpec extends ObjectBehavior
+{
+    public function it_is_initializable(): void
+    {
+        $this->shouldHaveType(MonthlySettlementPeriodResolver::class);
+    }
+
+    public function it_support_vendor_when_settlement_frequency_is_monthly(
```
Comment (senghe) [inline spec/Component/Settlement/PeriodStrategy/MonthlySettlementPeriodResolverSpec.php:25]: it support[s]

File: src/Component/Settlement/Cli/SettlementGenerateCommand.php [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    private function getSettlementDateRangeFromVendor(VendorInterface $vendor): array
+    {
+        /** @var AbstractSettlementPeriodResolverStrategy $settlementPeriodResolver */
+        foreach ($this->settlementPeriodResolvers as $settlementPeriodResolver) {
+            if ($settlementPeriodResolver->supports($vendor)) {
+                return $settlementPeriodResolver->resolve();
+            }
+        }
+
+        return [null, null];
```
Comment (senghe) [inline src/Component/Settlement/Cli/SettlementGenerateCommand.php:105]: We should throw an exception here

File: src/Component/Settlement/Resources/services/cli.xml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (senghe) [inline src/Component/Settlement/Resources/services/cli.xml:1]: Missing OpenSource header

File: tests/Integration/Cli/SettlementGenerateCommandTest.php [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        $kernel = self::bootKernel();
+        $application = new Application($kernel);
+        $command = $application->find('bitbag:settlement:generate');
+        $this->commandTester = new CommandTester($command);
+        $this->settlementRepository = self::getContainer()->get('open_marketplace.repository.settlement');
+        $this->vendorRepository = self::getContainer()->get('bitbag.open_marketplace.component.vendor.repository.vendor');
+        $this->channelRepository = self::getContainer()->get('sylius.repository.channel');
+    }
+
+    public function test_it_generates_settlements_for_all_vendors(): void
```
Comment (senghe) [inline tests/Integration/Cli/SettlementGenerateCommandTest.php:32]: I see there looks as it's a place for splitting assertions into private methods and reuse them.

File: tests/Integration/Cli/SettlementGenerateCommandTest.php [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        $this->assertSame($endDate->getTimestamp(), $settlementsVendorWeyland[0]->getEndDate()->getTimestamp());
+        $this->assertSame($channelUs, $settlementsVendorWeyland[0]->getChannel());
+        $this->assertSame(700, $settlementsVendorWeyland[1]->getTotalAmount());
+        $this->assertSame(100, $settlementsVendorWeyland[1]->getTotalCommissionAmount());
+        $this->assertSame($startDate->getTimestamp(), $settlementsVendorWeyland[1]->getStartDate()->getTimestamp());
+        $this->assertSame($endDate->getTimestamp(), $settlementsVendorWeyland[1]->getEndDate()->getTimestamp());
+        $this->assertSame($channelEu, $settlementsVendorWeyland[1]->getChannel());
+    }
+
+    private function getLastQuarterStartDate(): int
```
Comment (senghe) [inline tests/Integration/Cli/SettlementGenerateCommandTest.php:129]: I saw this method somewhere on the top - how about making it public and use here?

File: tests/Integration/Repository/SettlementRepositoryTest.php [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+{
+    public function setUp(): void
+    {
+        parent::setUp();
+        $this->repository = self::getContainer()->get('open_marketplace.repository.settlement');
+    }
+
+    public function test_it_finds_last_settlement_for_vendor(): void
+    {
+        $this->loadFixturesFromFile('SettlementRepositoryTest/test_it_find_last_settlement_for_vendor.yaml');
```
Comment (senghe) [inline tests/Integration/Repository/SettlementRepositoryTest.php:26]: it find(s)

File: tests/Integration/DataFixtures/ORM/SettlementRepositoryTest/test_it_find_last_settlement_for_vendor.yaml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    companyName: 'Wayne Enterprises, Inc.'
+    taxIdentifier: '1234567'
+    bankAccountNumber: 'PL97109024021972765458596357'
+    phoneNumber: '555123123'
+    slug: 'Wayne-Enterprises-Inc'
+    description: 'description'
+    commission: 10
+    commissionType: 'net'
+BitBag\OpenMarketplace\Component\Settlement\Entity\Settlement:
+  vendor_bruce_settlement_1:
```
Comment (senghe) [inline tests/Integration/DataFixtures/ORM/SettlementRepositoryTest/test_it_find_last_settlement_for_vendor.yaml:42]: How about having settlement for other vendor too here? The test would filter it out - and that's the check of vendor condition I think.

**Review by senghe — CHANGES_REQUESTED** [review]

File: spec/Component/Settlement/PeriodStrategy/MonthlySettlementPeriodResolverSpec.php (hunk 1 above)
Comment (beniaminblaszkiewicz) [reply spec/Component/Settlement/PeriodStrategy/MonthlySettlementPeriodResolverSpec.php]: fixed

File: src/Component/Settlement/Cli/SettlementGenerateCommand.php (hunk 2 above)
Comment (beniaminblaszkiewicz) [reply src/Component/Settlement/Cli/SettlementGenerateCommand.php]: fixed + added test

File: src/Component/Settlement/Resources/services/cli.xml (hunk 3 above)
Comment (beniaminblaszkiewicz) [reply src/Component/Settlement/Resources/services/cli.xml:1]: fixed

File: tests/Integration/Repository/SettlementRepositoryTest.php (hunk 6 above)
Comment (beniaminblaszkiewicz) [reply tests/Integration/Repository/SettlementRepositoryTest.php]: fixed

File: tests/Integration/DataFixtures/ORM/SettlementRepositoryTest/test_it_find_last_settlement_for_vendor.yaml (hunk 7 above)
Comment (beniaminblaszkiewicz) [reply tests/Integration/DataFixtures/ORM/SettlementRepositoryTest/test_it_find_last_settlement_for_vendor.yaml]: fixed

**Review by senghe — APPROVED** [review]

**Pull request merged**